### PR TITLE
Compatibility with Python 3.12.

### PR DESCRIPTION
**Pull request by veluca93**
Inspired by https://github.com/ioi-2024/cms/commit/0f044fd92c3d0b79e1588df015d6b0fdca930dee

### CODE REVIEW

File: cmscontrib/loaders/polygon.py [hunk 1]
```diff
@@ -145,10 +144,11 @@ def get_task(self, get_statement=True):
...
             logger.info("Found additional CMS options for task %s.", name)
-            with open(task_cms_conf_path, 'rb') as f:
-                task_cms_conf = imp.load_module('cms_conf', f,
-                                                task_cms_conf_path,
-                                                ('.py', 'r', imp.PY_SOURCE))
+            import importlib.util
+            spec = importlib.util.spec_from_file_location(
+                'cms_conf', task_cms_conf_path)
+            task_cms_conf = importlib.util.module_from_spec(spec)
+            spec.loader.exec_module(task_cms_conf)
```
Comment (wil93) [inline cmscontrib/loaders/polygon.py:151]: Have we tried running this on a polygon task by any chance? (TODO for the future: add a test with a polygon task)

File: requirements.txt [hunk 2]
```diff
@@ -1,20 +1,19 @@
...
 sqlalchemy>=1.3,<1.4  # http://docs.sqlalchemy.org/en/latest/changelog/index.html
 netifaces>=0.10,<0.11  # https://bitbucket.org/al45tair/netifaces/src/
-pycryptodomex>=3.6,<3.7  # https://github.com/Legrandin/pycryptodome/blob/master/Changelog.rst
+pycryptodomex==3.19.0  # https://github.com/Legrandin/pycryptodome/blob/master/Changelog.rst
 psutil>=5.5,<5.6  # https://github.com/giampaolo/psutil/blob/master/HISTORY.rst
-requests>=2.22,<2.23  # https://pypi.python.org/pypi/requests
-gevent==20.12.0  # http://www.gevent.org/changelog.html
-# Limit greenlet version for binary compatibility with gevent 20.12 wheels
-greenlet==1.0.0
-werkzeug>=0.16,<0.17  # https://github.com/pallets/werkzeug/blob/master/CHANGES
```
Comment (wil93) [inline requirements.txt:13]: Let's keep the comment here?

File: requirements.txt [hunk 3]
```diff
@@ -1,20 +1,19 @@
...
+requests==2.32.3  # https://pypi.python.org/pypi/requests
+gevent==23.9.0.post1  # http://www.gevent.org/changelog.html
+werkzeug<1.0
+backports.ssl-match-hostname==3.7.0.1
```
Comment (wil93) [inline requirements.txt:11]: What is this? 🤔

File: requirements.txt [hunk 4]
```diff
@@ -1,20 +1,19 @@
-# We are currently targeting the versions shipped in Ubuntu 20.04.
-
-tornado>=4.5,<4.6  # http://www.tornadoweb.org/en/stable/releases.html
-psycopg2>=2.8,<2.9  # http://initd.org/psycopg/articles/tag/release/
+# Fix tornado version to be sure that it will work with our patch
```
Comment (wil93) [inline requirements.txt:1]: This comment seems out of place, should we remove it?

File: cmscontrib/loaders/polygon.py (hunk 1 above)
Comment (veluca93) [reply cmscontrib/loaders/polygon.py:151]: No, and I don't think it's *too* important as this is a corner-case even in the polygon loader. Anyway, this is documented as equivalent in the python 3.12 docs.

File: requirements.txt (hunk 3 above)
Comment (veluca93) [reply requirements.txt]: Let's try if it works without it...

Comment (veluca93) [reply requirements.txt]: Tornado appears to import it but not depend on it (?)

File: cmscontrib/loaders/polygon.py (hunk 1 above)
Comment (wil93) [reply cmscontrib/loaders/polygon.py:151]: Indeed: https://docs.python.org/3.12/library/importlib.html#importing-a-source-file-directly

File: requirements.txt [hunk 5]
```diff
@@ -1,20 +1,18 @@
...
-requests>=2.22,<2.23  # https://pypi.python.org/pypi/requests
-gevent==20.12.0  # http://www.gevent.org/changelog.html
-# Limit greenlet version for binary compatibility with gevent 20.12 wheels
-greenlet==1.0.0
-werkzeug>=0.16,<0.17  # https://github.com/pallets/werkzeug/blob/master/CHANGES
+requests==2.32.3  # https://pypi.python.org/pypi/requests
+gevent==23.9.0.post1  # http://www.gevent.org/changelog.html
+werkzeug<1.0 # https://github.com/pallets/werkzeug/blob/master/CHANGES
+backports.ssl-match-hostname==3.7.0.1 # required by tornado<5.0
+greenlet>=3.0rc1
```
Comment (wil93) [inline requirements.txt:11]: nit: Let's keep the same order? i.e. keep greenlet and gevent together

**Review by wil93 — APPROVED** [review]

**Pull request merged**